Split out `receive_htlcs` from the forwarding pipeline

tnull opened the pull request:
This is another preparatory step for receiver-side delays that we split out to err on the side of smaller, more reviewable PRs, especially since these changes are a nice cleanup in any case, IMO.

Previously, we'd store receiving HTLCs side-by-side with HTLCs forwards in the `forwards_htlcs` map under SCID 0.
Here, we opt to split out a separate `receive_htlcs` field which cleans up the logic, also omitting the 0 magic value.

Moreover, some of the tests manipulating `forward_htlcs` were previously just iterating, but not actually checking whether the expected entries were present and they were actually changed. Here we make these test cases stricter to ensure they'd able to catch any unwanted behavior we'd introduce while introducing  `receive_htlcs`.

tnull committed:
Move receiving HTLCs to a `receive_htlcs` field

Previously, we'd store receiving HTLCs side-by-side with HTLCs forwards
in the `forwards_htlcs` field under SCID 0.

Here, we opt to split out a separate `receive_htlcs` field, also
omitting the 0 magic value.

tnull committed:
Make tests touching `forward_htlcs` more strict

Previously, some of the tests manipulating `forward_htlcs` were just
iterating, but not actually checking whether the expected entries were
present and they were actually changed. Here we make these test cases
more strict.

martinsaposnic commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6402,22 +6414,18 @@ where
 
+		let mut receive_htlcs = Vec::new();
+		mem::swap(&mut receive_htlcs, &mut self.receive_htlcs.lock().unwrap());
+		self.process_receive_htlcs(receive_htlcs, &mut new_events, &mut failed_forwards);
```
if there are 0 forward_htlcs but non-0 receive_htlcs, then `should_persist` won't be set to `NotifyOption::DoPersist;`.  (it's only set here https://github.com/lightningdevkit/rust-lightning/pull/3973/commits/2f5e9409b37bb0fcf05a5ecae6f6eb37c0a5daf6#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cR6416 when iterating over the forward_htlcs)

maybe that could cause a persistence problem?

valentinewallace reviewed (commented):
Basically LGTM!

Could you confirm one thing for me: if we fail to receive an HTLC that's destined for us, it will then be put into the `forward_htlcs` map keyed with the scid of the previous hop? That's how I'm reading the behavior of `fail_htlc_backwards_internal` atm. If that's the case, may want to note that on the `forward_htlcs` field.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -16791,6 +16854,7 @@ where
 			pending_intercepted_htlcs: Mutex::new(pending_intercepted_htlcs.unwrap()),
 
 			forward_htlcs: Mutex::new(forward_htlcs),
+			receive_htlcs: Mutex::new(receive_htlcs),
```
Earlier in this method, we remove pending payments that are not present in the monitors, and ISTM we should be doing this for `receive_htlcs` now as well

tnull committed:
Fix `test_overshoot_final_cltv`

Previously, the `pending_forwards` in this test would simply be empty,
having the test not run any of the subsequent logic.

Here we add a necessary `process_pending_update_add_htlcs()` step, and
will move to a more strict test logic in the following commits.

tnull committed:
f Use other match pattern here already

.. to keep changes in the following commits minimal.

tnull committed:
f Add comment that scid == 0 means trampoline now
